From 18d12f1792618704c9b43ef3ce6e9d10dd55834f Mon Sep 17 00:00:00 2001
From: Julian Bouzas <julian.bouzas@collabora.com>
Date: Tue, 13 Oct 2026 05:48:07 -0400
Subject: [PATCH] spa: bluez5: use a unique serial for device and transport
 properties

This patch avoids using a raw pointer as ID for 'api.bluez5.transport' and
'api.bluez5.device' properties and instead uses a unique serial. This allows
impl_init() to check if the given serial represents a valid transport or device,
even if it was already destroyed.
---
 spa/plugins/bluez5/bluez5-dbus.c   | 43 +++++++++++++++++++++++++++++-
 spa/plugins/bluez5/bluez5-device.c | 12 +++++----
 spa/plugins/bluez5/defs.h          |  4 +++
 spa/plugins/bluez5/media-sink.c    | 10 ++++---
 spa/plugins/bluez5/media-source.c  | 10 ++++---
 5 files changed, 65 insertions(+), 14 deletions(-)

diff --git a/spa/plugins/bluez5/bluez5-dbus.c b/spa/plugins/bluez5/bluez5-dbus.c
index 34b39a98f..e4a6e1d67 100644
--- a/spa/plugins/bluez5/bluez5-dbus.c
+++ b/spa/plugins/bluez5/bluez5-dbus.c
@@ -86,6 +86,8 @@ struct spa_bt_monitor {
 	struct spa_handle handle;
 	struct spa_device device;
 
+	struct spa_list link;
+
 	struct spa_log *log;
 	struct spa_loop *main_loop;
 	struct spa_loop *data_loop;
@@ -1769,6 +1771,27 @@ static uint32_t adapter_connectable_profiles(struct spa_bt_adapter *adapter)
 	return mask;
 }
 
+/* Used to get a unique serial for every object */
+static uint32_t serial_counter;
+
+/*
+ * List of all monitors, so that device and node factories can find the
+ * device or transport by serial.
+ */
+static struct spa_list monitor_list = SPA_LIST_INIT(&monitor_list);
+
+struct spa_bt_device *spa_bt_device_find_by_serial(uint32_t serial)
+{
+	struct spa_bt_monitor *monitor;
+	struct spa_bt_device *d;
+
+	spa_list_for_each(monitor, &monitor_list, link)
+		spa_list_for_each(d, &monitor->device_list, link)
+			if (d->serial == serial)
+				return d;
+	return NULL;
+}
+
 struct spa_bt_device *spa_bt_device_find(struct spa_bt_monitor *monitor, const char *path)
 {
 	struct spa_bt_device *d;
@@ -1809,6 +1832,7 @@ static struct spa_bt_device *device_create(struct spa_bt_monitor *monitor, const
 		return NULL;
 
 	d->id = monitor->id++;
+	d->serial = serial_counter++;
 	d->monitor = monitor;
 	d->path = strdup(path);
 	d->battery_path = battery_get_name(d->path);
@@ -2028,7 +2052,7 @@ static void emit_device_info(struct spa_bt_monitor *monitor,
 	items[n_items++] = SPA_DICT_ITEM_INIT(SPA_KEY_API_BLUEZ5_ICON, device->icon);
 	items[n_items++] = SPA_DICT_ITEM_INIT(SPA_KEY_API_BLUEZ5_PATH, device->path);
 	items[n_items++] = SPA_DICT_ITEM_INIT(SPA_KEY_API_BLUEZ5_ADDRESS, device->address);
-	snprintf(dev, sizeof(dev), "pointer:%p", device);
+	snprintf(dev, sizeof(dev), "%u", device->serial);
 	items[n_items++] = SPA_DICT_ITEM_INIT(SPA_KEY_API_BLUEZ5_DEVICE, dev);
 	snprintf(class, sizeof(class), "0x%06x", device->bluetooth_class);
 	items[n_items++] = SPA_DICT_ITEM_INIT(SPA_KEY_API_BLUEZ5_CLASS, class);
@@ -3233,6 +3257,18 @@ static void remote_endpoint_free(struct spa_bt_remote_endpoint *remote_endpoint)
 	free(remote_endpoint);
 }
 
+struct spa_bt_transport *spa_bt_transport_find_by_serial(uint32_t serial)
+{
+	struct spa_bt_monitor *monitor;
+	struct spa_bt_transport *t;
+
+	spa_list_for_each(monitor, &monitor_list, link)
+		spa_list_for_each(t, &monitor->transport_list, link)
+			if (t->serial == serial)
+				return t;
+	return NULL;
+}
+
 struct spa_bt_transport *spa_bt_transport_find(struct spa_bt_monitor *monitor, const char *path)
 {
 	struct spa_bt_transport *t;
@@ -3263,6 +3299,7 @@ struct spa_bt_transport *spa_bt_transport_create(struct spa_bt_monitor *monitor,
 	if (t == NULL)
 		return NULL;
 
+	t->serial = serial_counter++;
 	t->acquire_refcount = 0;
 	t->monitor = monitor;
 	t->path = path;
@@ -6950,6 +6987,8 @@ static int impl_clear(struct spa_handle *handle)
 
 	monitor = (struct spa_bt_monitor *) handle;
 
+	spa_list_remove(&monitor->link);
+
 	/*
 	 * We don't call BlueZ API unregister methods here, since BlueZ generally does the
 	 * unregistration when the DBus connection is closed below.  We'll unregister DBus
@@ -7609,6 +7648,8 @@ impl_init(const struct spa_handle_factory *factory,
 	this->backends[BACKEND_OFONO] = backend_ofono_new(this, this->conn, info, this->quirks, support, n_support);
 	this->backends[BACKEND_HSPHFPD] = backend_hsphfpd_new(this, this->conn, info, this->quirks, support, n_support);
 
+	spa_list_append(&monitor_list, &this->link);
+
 	return 0;
 
 fail:
diff --git a/spa/plugins/bluez5/bluez5-device.c b/spa/plugins/bluez5/bluez5-device.c
index 350b8b591..fc5b311d9 100644
--- a/spa/plugins/bluez5/bluez5-device.c
+++ b/spa/plugins/bluez5/bluez5-device.c
@@ -730,7 +730,7 @@ static void emit_node(struct impl *this, struct spa_bt_transport *t,
 	else
 		in_device_set = this->device_set.source_enabled;
 
-	snprintf(transport, sizeof(transport), "pointer:%p", t);
+	snprintf(transport, sizeof(transport), "%u", t->serial);
 	items[0] = SPA_DICT_ITEM_INIT(SPA_KEY_API_BLUEZ5_TRANSPORT, transport);
 	items[1] = SPA_DICT_ITEM_INIT(SPA_KEY_API_BLUEZ5_PROFILE, spa_bt_profile_name(t->profile));
 	items[2] = SPA_DICT_ITEM_INIT(SPA_KEY_API_BLUEZ5_CODEC, get_codec_name(t, a2dp_duplex));
@@ -3536,6 +3536,7 @@ impl_init(const struct spa_handle_factory *factory,
 {
 	struct impl *this;
 	const char *str;
+	uint32_t serial;
 	unsigned int i;
 
 	spa_return_val_if_fail(factory != NULL, -EINVAL);
@@ -3551,12 +3552,13 @@ impl_init(const struct spa_handle_factory *factory,
 
 	spa_log_topic_init(this->log, &log_topic);
 
-	if (info && (str = spa_dict_lookup(info, SPA_KEY_API_BLUEZ5_DEVICE)))
-		sscanf(str, "pointer:%p", &this->bt_dev);
+	if (info && (str = spa_dict_lookup(info, SPA_KEY_API_BLUEZ5_DEVICE)) &&
+	    spa_atou32(str, &serial, 10))
+		this->bt_dev = spa_bt_device_find_by_serial(serial);
 
 	if (this->bt_dev == NULL) {
-		spa_log_error(this->log, "a device is needed");
-		return -EINVAL;
+		spa_log_debug(this->log, "could not find device by serial");
+		return -ESTALE;
 	}
 
 	if (info) {
diff --git a/spa/plugins/bluez5/defs.h b/spa/plugins/bluez5/defs.h
index c9ad50344..6ecc08749 100644
--- a/spa/plugins/bluez5/defs.h
+++ b/spa/plugins/bluez5/defs.h
@@ -528,6 +528,7 @@ struct spa_bt_device {
 	struct spa_bt_monitor *monitor;
 	struct spa_bt_adapter *adapter;
 	uint32_t id;
+	uint32_t serial;
 	char *path;
 	char *alias;
 	char *address;
@@ -579,6 +580,7 @@ struct spa_bt_device {
 	bool disable_dummy_call;
 };
 
+struct spa_bt_device *spa_bt_device_find_by_serial(uint32_t serial);
 struct spa_bt_device *spa_bt_device_find(struct spa_bt_monitor *monitor, const char *path);
 struct spa_bt_device *spa_bt_device_find_by_address(struct spa_bt_monitor *monitor, const char *remote_address, const char *local_address);
 int spa_bt_device_add_profile(struct spa_bt_device *device, enum spa_bt_profile profile);
@@ -670,6 +672,7 @@ struct spa_bt_transport_volume {
 
 struct spa_bt_transport {
 	struct spa_list link;
+	uint32_t serial;
 	struct spa_bt_monitor *monitor;
 	struct spa_bt_backend *backend;
 	char *path;
@@ -727,6 +730,7 @@ struct spa_bt_transport {
 struct spa_bt_transport *spa_bt_transport_create(struct spa_bt_monitor *monitor, char *path, size_t extra);
 void spa_bt_transport_free(struct spa_bt_transport *transport);
 void spa_bt_transport_set_state(struct spa_bt_transport *transport, enum spa_bt_transport_state state);
+struct spa_bt_transport *spa_bt_transport_find_by_serial(uint32_t serial);
 struct spa_bt_transport *spa_bt_transport_find(struct spa_bt_monitor *monitor, const char *path);
 struct spa_bt_transport *spa_bt_transport_find_full(struct spa_bt_monitor *monitor,
                                                     bool (*callback) (struct spa_bt_transport *t, const void *data),
diff --git a/spa/plugins/bluez5/media-sink.c b/spa/plugins/bluez5/media-sink.c
index 1903c7965..aaf19ea17 100644
--- a/spa/plugins/bluez5/media-sink.c
+++ b/spa/plugins/bluez5/media-sink.c
@@ -2566,6 +2566,7 @@ impl_init(const struct spa_handle_factory *factory,
 	struct impl *this;
 	struct port *port;
 	const char *str;
+	uint32_t serial;
 
 	spa_return_val_if_fail(factory != NULL, -EINVAL);
 	spa_return_val_if_fail(handle != NULL, -EINVAL);
@@ -2647,12 +2648,13 @@ impl_init(const struct spa_handle_factory *factory,
 	if (info && (str = spa_dict_lookup(info, "bluez5.debug.iso-mono")) != NULL)
 		this->iso_debug_mono = spa_atob(str);
 
-	if (info && (str = spa_dict_lookup(info, SPA_KEY_API_BLUEZ5_TRANSPORT)))
-		sscanf(str, "pointer:%p", &this->transport);
+	if (info && (str = spa_dict_lookup(info, SPA_KEY_API_BLUEZ5_TRANSPORT)) &&
+	    spa_atou32(str, &serial, 10))
+		this->transport = spa_bt_transport_find_by_serial(serial);
 
 	if (this->transport == NULL) {
-		spa_log_error(this->log, "a transport is needed");
-		return -EINVAL;
+		spa_log_debug(this->log, "could not find transport by serial");
+		return -ESTALE;
 	}
 	if (this->transport->media_codec == NULL) {
 		spa_log_error(this->log, "a transport codec is needed");
diff --git a/spa/plugins/bluez5/media-source.c b/spa/plugins/bluez5/media-source.c
index d10250181..e09a3dcc8 100644
--- a/spa/plugins/bluez5/media-source.c
+++ b/spa/plugins/bluez5/media-source.c
@@ -2147,6 +2147,7 @@ impl_init(const struct spa_handle_factory *factory,
 	struct impl *this;
 	struct port *port;
 	const char *str;
+	uint32_t serial;
 
 	spa_return_val_if_fail(factory != NULL, -EINVAL);
 	spa_return_val_if_fail(handle != NULL, -EINVAL);
@@ -2225,11 +2226,12 @@ impl_init(const struct spa_handle_factory *factory,
 
 	this->quantum_limit = 8192;
 
-	if (info && (str = spa_dict_lookup(info, SPA_KEY_API_BLUEZ5_TRANSPORT)) != NULL)
-		sscanf(str, "pointer:%p", &this->transport);
+	if (info && (str = spa_dict_lookup(info, SPA_KEY_API_BLUEZ5_TRANSPORT)) != NULL &&
+	    spa_atou32(str, &serial, 10))
+		this->transport = spa_bt_transport_find_by_serial(serial);
 	if (this->transport == NULL) {
-		spa_log_error(this->log, "a transport is needed");
-		return -EINVAL;
+		spa_log_debug(this->log, "could not find transport by serial");
+		return -ESTALE;
 	}
 	if (this->transport->media_codec == NULL) {
 		spa_log_error(this->log, "a transport codec is needed");
-- 
2.56.0

